hnn_core/network.py

```diff
@@ -431,6 +434,71 @@ def update_types(self, gid_dict):
             spike_types += [list(spike_types_trial)]
         self._types = spike_types
 
+    def mean_rates(self, tstart, tstop, gid_dict, mean_type='all'):
+        """Mean spike rates (Hz) by cell type.
+
+        Parameters
+        ----------
+        tstart : int | float | None
+            Value defining the start time of all trials.
+        tstop : int | float | None
+            Value defining the stop time of all trials.
+        gid_dict : dict of lists or range objects
+            Dictionary with keys 'evprox1', 'evdist1' etc.
+            containing the range of Cell or input IDs of different
+            cell or input types.
+        mean_type : str
+            'all' : Average over trials and cells
+                Returns mean rate for cell types
+            'trial' : Average over cell types
+                Returns trial mean rate for cell types
+            'cell' : Average over individual cells
+                Returns trial mean rate for individual cells
+
+        Returns
+        -------
+        spike_rate : dict
+            Dictionary with keys 'L5_pyramidal', 'L5_basket', etc.
+        """
+        cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
+        spike_rates = dict()
+
+        if mean_type not in ['all', 'trial', 'cell']:
+            raise ValueError("Invalid mean_type. Valid arguments include "
+                             "'all', 'trial', or 'cell'.")
+
+        # Validate tstart, tstop
+        if not isinstance(tstart, (int, float)) or not isinstance(
+                tstop, (int, float)):
+            raise ValueError('tstart and tstop must be of type int or float')
+        elif tstop <= tstart:
+            raise ValueError('tstop must be greater than tstart')
+
+        for cell_type in cell_types:
+            cell_type_gids = np.array(gid_dict[cell_type])
+            gid_spike_rate = np.zeros((len(self._times), len(cell_type_gids)))
```

Comment:
Reading the code it was not obvious to me that `len(self._times)` is actually number of trials. I would add one line above to make it obvious
```suggestion
            n_trials, n_cells = len(self._times), len(cell_type_gids)
            gid_spike_rate = np.zeros((n_trials, n_cells))
```